src/haliax/_src/state_dict.py

```diff
@@ -0,0 +1,491 @@
+                # via pjit. If we can't find one, we'll just fully replicate since it probably isn't that big.
+                # TODO: ensure that this mesh arranges devices correctly
+                # (jax seems to do this internally itself, so we should be fine?)
+                process_mesh = Mesh(np.array(jax.devices()).reshape((jax.process_count(), -1)), ("process", "device"))
+                # now we need to find an axis along which we can shard the array.
+                # for this, we need to find an axis s.t. size(axis) % local_devices == 0
+
+                try:
+                    axis_to_shard = index_where(
+                        lambda axis_size: axis_size % process_mesh.devices.size == 0, arr.shape
+                    )
+                except ValueError:
+                    return np.array(arr)
+
+                shardings = [None if i != axis_to_shard else "device" for i in range(len(arr.shape))]
+                sharding = NamedSharding(process_mesh, PartitionSpec(*shardings))
+                out = jax.jit(lambda x: x, out_shardings=sharding)(arr)
+                return np.array(out)
+
+        # need to make sure the model is on *this machine* and *this machine's CPU* before saving
+        model = jax.tree_util.tree_map(lambda arr: get_to_cpu(arr), model)
+        # TODO: it would be nice if safetensors supported an iterator or something so we could do the allgather one at a time
+        state_dict = to_state_dict(model, prefix=prefix)
+        return state_dict
+
+
+_GLOBAL_SAVE_COUNT = 0
+
+
+def save_state_dict(state_dict: StateDict, path):
+    """
+    Save a model's state dict to a file, bringing all tensors to the CPU first and then converting to numpy.
+    This will save using safetensors format
+    """
+    state_dict = {k: v for k, v in state_dict.items() if v is not None}
+    # now that we've moved the model to the CPU, we don't need to do this on all processes
+    if jax.process_index() == 0:
+        # the "pt" is a lie but it doesn't seem to actually matter and HF demands it
+        safetensors.numpy.save_file(state_dict, path, metadata={"format": "pt"})
+    global _GLOBAL_SAVE_COUNT
+    sync_global_devices(f"save_state_dict {_GLOBAL_SAVE_COUNT}")
+    _GLOBAL_SAVE_COUNT += 1
+
+
+def load_state_dict(path):
+    """
+    Load a model's state dict from a file, bringing all tensors to the CPU first and then converting to numpy.
+    This will load using safetensors format
+    """
+    state_dict = safetensors.numpy.load_file(path)
+    return state_dict
+
+
+def stack_state_dict(state_dict: StateDict, prefix: Optional[str] = None) -> StateDict:
+    """
+    Stack all keys matching prefix in a new state dict, returning a state dict that has all keys matching
+    prefix stacked, but otherwise the same.
+
+    Stacked in this case means roughly "compatible with a torch.nn.Sequential", which means that the
```

Comment:
This comment is the same for `stacked` and `unstacked` below. Maybe:

"The unstacked format is compatible with `torch.nn.Sequential`, with keys of the form (...). The stacked/vectorized format is required for `haliax.nn.Stacked` and vectorizes all such tensors into a single shared key.".

Reply:
oops, thanks